tree.go

```diff
@@ -513,30 +513,52 @@ func (n *InternalNode) toHashedNode() *HashedNode {
 	return &HashedNode{commitment: comm[:]}
 }
 
-func (n *InternalNode) Delete(key []byte, resolver NodeResolverFn) error {
+func (n *InternalNode) Delete(key []byte, resolver NodeResolverFn) (error, bool) {
 	nChild := offset2key(key, n.depth)
 	switch child := n.children[nChild].(type) {
 	case Empty:
-		return nil
+		return nil, false
 	case *HashedNode:
 		if resolver == nil {
-			return errDeleteHash
+			return errDeleteHash, false
 		}
 		comm := child.commitment
 		payload, err := resolver(comm)
 		if err != nil {
-			return err
+			return err, false
 		}
 		// deserialize the payload and set it as the child
 		c, err := ParseNode(payload, n.depth+1, comm)
 		if err != nil {
-			return err
+			return err, false
 		}
 		n.children[nChild] = c
 		return n.Delete(key, resolver)
 	default:
 		n.cowChild(nChild)
-		return child.Delete(key, resolver)
+		err, delete := child.Delete(key, resolver)
+		if err != nil {
+			return err, false
+		}
+
+		// delete the entire child if instructed to by
+		// the recursive algorightm.
+		if delete {
+			n.children[nChild] = Empty{}
```

Comment:
This will be an obvious comment, but we're changing the spec, right?

Reply:
we're not changing the spec in the sense that going forward along the chain, no deletion is possible. But in the case of reorgs, it must be possible to delete values that were inserted in the tree while executing the reorg'ed blocks.

Reply:
So to be completely clear here: instead of calling `Delete(key)` in `(*VerkleTrie).Delete`, one should call `Update(key, 0)` to overwrite it with 0s. Geth is responsible for following the spec, this library implements proper deletion as it will be used for the path-based approach to implement deep reorgs.

Reply:
Ah, OK, so it's an API usage change from the client side compared to before. Makes sense.